_posts/2016-08-31-continuous-delivery-databases.md

```diff
@@ -0,0 +1,57 @@
+* Es muss weitgehend möglich sein, die Migration von persistenten Umgebungen (Datenbank bleibt beim Neudeployment der Umgebung bestehen) vorher zu testen. Dies betrifft also insbesondere PreProduction und Production
+* Die Migration muss automatisiert durch Skripte erfolgen, damit sich durch manuelle Eingriffe keine Fehler einschleichen und der Aufwand gering bleibt.
+* Wir müssen auch "Basisdaten" verwalten können. Dies umfasst z.B. Referenztabellen (z.B. Ländercodes).
+
+## Kompromisse
+
+Häufig müssen bei Qualitätskritieren Kompromisse eingeangen werden. Ein für uns sehr wichtiger Kompromiss ist der zwischen Qualität der Datenmigration und Dauer für bestimmte Schritte in der Delivery Pipeline.
+
+Da wir für alle Umgebungen ein gleichartiges Vorgehen nutzen, bedeutet dies für einige Umgebungen, insbesondere bei denen es auch eine ephemerale Datenbank gibt, dass das Erzeugen des Schemas länger dauert als wenn man die Datenbank durch ein einziges Skript auf den aktuell gültigen Zustand bringt und nicht inkrementelle Änderungen anwendet (die ja mit der Zeit immer länger werden).
+Dies hat Auswirkungen z.B. auf die Dauer von E2E Tests. Um dies zu kompensieren, kann die Nutzung von mehr Ressourcen notwendig sein. Die Zusatzkosten sind aber unserer Erfahrung nach geringer als die Kosten aufgrund doppelter Pflege und
+Nacharbeiten aufgrund schlechter Qualität.  
+
+## Lösungsmöglichkeit
+
+Wir verwenden eine komplett auf Docker basierende Continuous Delivery Pipeline. Dies bedeutet, dass wir ab einem bestimmten Schritt nicht
+mehr mit den jar / war Artefakten arbeiten, sondern dirket mit den erzeugten Docker Images.
+
+Unter diesen Randbedingungen sieht unsere Lösung in etwa wie folgt aus:
+
+* Wir verwenden Liquibase, da das Tool datenbankunabhängig ist und wir damit einfach MySQL, unterschiedliche Kontexte und Basisdaten verwalten können.
+* Jede Applikation / Service enthält ein Modul "dbmigrations", welches alle notwendigen Liquibase Migrationen enthält (als SQL Dateien).
+* In einem bestimmten Schritt erzeugen wir die entsprechenden Docker Images, unter anderem dabei auch das Liquibase Image. Dazu holen wir die dbmigrations Module aus dem Maven Repository. Dieses Image beinhaltet ein Skript, welches die DB Migrations dann auch auf eine Zieldatenbank anwenden kann.
+* Es gibt einen weiteren Pipelineschritt, der frühzeitig feststellen soll, ob die Migrations "fehlerfrei" sind und auf eine bestehende Datenbank angewendet werden können. Dazu holen wir das Liquibase DB Migrations Image, welches aktuell auf der UAT Umgebung deployed wurde. Dieses wenden wir auf eine "leere" Datenbank an. Anschließend wird das aktuell gebaute Liqubiase Image auf dieser Datenbank angewendet. Dies setzt voraus, dass niemand manuell Änderungen an der Datenbank durchführt. Ansonsten sollte man immer das DDL direkt aus der Zieldatenbank (UAT, Prod) erzeugen, gegen welche man testen will.
```

Comment:
Selbes gilt hier: `Datenbankmigrationen`

Warum steht `"fehlerfrei"` in Anführungszeichen. Das bedarf Erklärung, ansonsten weg damit.